src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -607,34 +607,82 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
             }
 
             ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
             return default;
         }
 
-        internal virtual void WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options, ref WriteStack state)
+        internal virtual T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            Debug.Assert(reader.TokenType == JsonTokenType.PropertyName);
+
+            long originalBytesConsumed = reader.BytesConsumed;
+            T result = ReadAsPropertyName(ref reader, typeToConvert, options);
+            if (reader.BytesConsumed != originalBytesConsumed)
+            {
+                ThrowHelper.ThrowJsonException_SerializationConverterRead(this);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Writes a value as a JSON property name.
+        /// </summary>
+        /// <param name="writer">The <see cref="Utf8JsonWriter"/> to write to.</param>
+        /// <param name="value">The value to convert. Note that the value of <seealso cref="HandleNull"/> determines if the converter handles <see langword="null" /> values.</param>
+        /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
+        public virtual void WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
         {
             if (!IsInternalConverter && options.TryGetDefaultSimpleConverter(TypeToConvert, out JsonConverter? defaultConverter))
             {
                 // .NET 5 backward compatibility: hardcode the default converter for primitive key serialization.
                 Debug.Assert(defaultConverter.IsInternalConverter && defaultConverter is JsonConverter<T>);
-                ((JsonConverter<T>)defaultConverter).WriteAsPropertyName(writer, value, options, ref state);
+                ((JsonConverter<T>)defaultConverter).WriteAsPropertyNameCore(writer, value, options, isWritingExtensionDataProperty: false);
                 return;
             }
 
             ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
         }
 
-        internal sealed override void WriteAsPropertyNameAsObject(Utf8JsonWriter writer, object value, JsonSerializerOptions options, ref WriteStack state)
-            => WriteAsPropertyName(writer, (T)value, options, ref state);
+        internal virtual void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, JsonSerializerOptions options, bool isWritingExtensionDataProperty)
+        {
+            if (isWritingExtensionDataProperty)
+            {
+                // Since we can't pass isWritingExtensionDataProperty information to the public WriteAsPropertyName method,
+                // err on the side of backward compatibility and hardcode the default string property name write semantics.
```

Comment:
It's not really about backwards compat - users don't want extension data transformed in any way since it is often used to round-trip unused or extra data as-is.

Reply:
Thanks, that's good feedback. My interpretation of the extension data feature is that it was dictionary serialization where entries were simply being flattened in the parent dictionary. As such I expected custom key serialization to be a reasonable feature to be supported. I'll update the remark.